nexus/db-queries/src/db/datastore/ip_pool.rs

```diff
@@ -256,6 +419,311 @@ impl DataStore {
+            .limit(1)
+            .load_async::<ExternalIp>(
+                &*self.pool_connection_authorized(opctx).await?,
+            )
+            .await
+            .map_err(|e| {
+                Error::internal_error(&format!(
+                    "error checking for outstanding IPs before deleting IP pool association to resource: {:?}",
+                    e
+                ))
+            })?;
+
+        if !existing_ips.is_empty() {
+            return Err(Error::invalid_request(
+                "IP addresses from this pool are in use in the linked silo",
+            ));
+        }
+
+        Ok(())
+    }
+
+    /// Floating IPs are associated with a silo through a project, so this one
+    /// is a little simpler than ephemeral. We join IP -> Project -> Silo.
+    async fn ensure_no_floating_ips_outstanding(
+        &self,
+        opctx: &OpContext,
+        association: &IpPoolResourceDelete,
+    ) -> Result<(), Error> {
+        use db::schema::external_ip;
+        use db::schema::project;
+        opctx
+            .authorize(authz::Action::CreateChild, &authz::IP_POOL_LIST)
+            .await?;
+
+        let existing_ips = external_ip::table
+            .inner_join(project::table.on(external_ip::project_id.eq(project::id.nullable())))
+            .filter(external_ip::is_service.eq(false))
+            .filter(external_ip::time_deleted.is_null())
+            // all floating IPs have a project
+            .filter(external_ip::project_id.is_not_null())
+            .filter(external_ip::ip_pool_id.eq(association.ip_pool_id))
+            .filter(external_ip::kind.eq(IpKind::Floating))
+            // we have to join through IPs to projects to get the silo ID
+            .filter(project::silo_id.eq(association.resource_id))
+            .filter(project::time_deleted.is_null())
+            .select(ExternalIp::as_select())
+            .limit(1)
+            .load_async::<ExternalIp>(
+                &*self.pool_connection_authorized(opctx).await?,
+            )
+            .await
+            .map_err(|e| {
+                Error::internal_error(&format!(
+                    "error checking for outstanding IPs before deleting IP pool association to resource: {:?}",
+                    e
+                ))
+            })?;
+
+        if !existing_ips.is_empty() {
+            return Err(Error::invalid_request(
+                "IP addresses from this pool are in use in the linked silo",
+            ));
+        }
+
+        Ok(())
+    }
+
+    /// Delete IP pool assocation with resource unless there are outstanding
+    /// IPs allocated from the pool in the associated silo
+    pub async fn ip_pool_dissociate_resource(
+        &self,
+        opctx: &OpContext,
+        association: &IpPoolResourceDelete,
+    ) -> DeleteResult {
+        use db::schema::ip_pool_resource;
+        opctx
+            .authorize(authz::Action::CreateChild, &authz::IP_POOL_LIST)
+            .await?;
```

Comment:
This doesn't feel like the right check. I think again, either creating an authz type for the relationship or checking modify on one or both of the resources being changed.